content/source/docs/extend/debugging.html.md

```diff
@@ -0,0 +1,179 @@
+---
+layout: "extend"
+page_title: "Extending Terraform: Debugging Providers"
+sidebar_current: "docs-extend-debugging-providers"
+description: |-
+  How to debug providers.
+---
+
+# Debugging Providers
+
+Terraform's providers are complicated programs in their own right, and
+sometimes they don't end up working the way you expect them to. In those
+situations, it's not always obvious how to get information about what's
+actually happening inside a provider or why things are behaving the way they
+do.
+
+This guide documents a few different ways to access more information about the
+runtime operations of Terraform providers. It is intended for Terraform
+provider developers, though sufficiently advanced users may also be able to use
+it.
+
+There are two available approaches to debugging Terraform providers. We'll talk
+about each of them separately and in turn.
+
+## Log-Based Debugging
+
+Log-based debugging is a method of using logging calls to record what is
+happening in a provider as it happens, and then examining that record to piece
+together what happened after the fact. Usefully, end users can capture logs and
+share them with developers, allowing developers to sometimes debug providers
+without needing access to the environment or configuration. Unfortunately,
+developers need to think ahead about what information they'll need when
+debugging, because the log lines need to be injected into the provider when
+it's built and cannot be inserted into the binary after it has been built. This
+makes it harder to ask new questions when doing log-based debugging, as the log
+line needs to be added and the provider needs to be recompiled before new
+information can be surfaced.
+
+### Inserting Log Lines Into A Provider
+
+Terraform providers all log to `stderr`. Any output to `stderr` will be passed
+over the gRPC protocol to Terraform, which will then write it to its
+destination.
+
+~> **NOTE:** Don't write to `stdout` on any Terraform providers! It can cause
```

Comment:
Yellow boxes -> "important" instead of "note".

```suggestion
~> **Important:** Don't write to `stdout` on any Terraform providers! It can cause
```